pkg/activator/prober.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package activator
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"go.uber.org/zap"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	maxRetry              = 60
+	defaultPeriodSeconds  = int32(1 * time.Second)
+	defaultTimeoutSeconds = int32(1 * time.Second)
+)
+
+func verifyRevisionRoutability(revision *v1alpha1.Revision, endpoint *Endpoint, logger *zap.SugaredLogger) {
+	// Proceed only if user has a HTTPGet readinessProbe defined
+	if revision.Spec.Container.ReadinessProbe == nil ||
+		revision.Spec.Container.ReadinessProbe.HTTPGet == nil {
+		endpoint.Verified = Unknown
+		return
+	}
+
+	endpoint.Verified = Fail
+	probe := createHttpGetProbe(revision, *endpoint)
+
+	// Number of seconds after the readiness probes are initiated
+	time.Sleep(time.Second * int32ToDuration(probe.InitialDelaySeconds))
```

Comment:
We did have some hesitation in using it. But the thinking was that a user directive to apply an initial delay should be respected. The default initial delay is 0 seconds.

As an example, assume for an app, the actual initial delay is longer than 60 seconds; then the endpoint verification would fail if the default retry interval of 1 second and max retry limit of 60 is used.